mlx/backend/metal/resident.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright © 2024 Apple Inc.
+
+#include "mlx/backend/metal/resident.h"
+#include "mlx/backend/metal/metal_impl.h"
+
+namespace mlx::core::metal {
+
+// TODO maybe worth including tvos / visionos
+#define supported __builtin_available(macOS 15, iOS 18, *)
+
+ResidencySet::ResidencySet(MTL::Device* d) {
+  if (supported) {
```

Comment:
Couple things here:
- This is a runtime check because otherwise we can't use this feature in MLX Swift binaries without breaking macOS<15 support.
- It has to be `if (__builtin_available(...))` so that's why it's with the preprocessor instead of say a `bool` or something like that.

Reply:
Does it also have to be `if (supported)`? I am asking because

```cpp
if (!supported) {
  return;
}
```

could be a tiny bit simpler.

Reply:
It's crazy but it actually does. There might be a `__not_builtin_available` or something

Reply:
I can't find anything like that sadly.